rfcs/001-foundational-architecture.md

```diff
@@ -0,0 +1,564 @@
+  filesystem.writeFile(resources)
+}
+
+// The config can also be used to export
+// variables or objects.
+export const metadata = {
+  language: "DE",
+}
+
+```
+
+One (the?) drawback of JS as config is exploit vulnerability. The JS config could contain malicious code that would be executed by inlang components. An example exploit: An attacker could steal user authentification information by writing malicious code in the config that reads authentification information from the editor. JS as config would require sandboxing to a certain degree to eliminate exploit vulnerability.
+
+#### Proposal
+
+Leveraging JavaScript, or any programming language allows for tremendous flexibility and therefore unopinionated workflows. Flexibility is required: codebases differ, approaches to localize software differ, syntaxes differ, and last but not least workflows differ. JavaScript as config could even be used to adjust the business logic of components:
+
+```js
+// pseudocode that illustrates the possbility to
+// adjust how messages are machine translated.
+export async function onMachineTranslate(message) {
+  // or DeepL, Yandex, etc.
+  const googleTranslate = await import("google-translate");
+  return googleTranslate(message);
+}
+
+// pseudocode that illustrates the possiblity to adjust
+// the business logic of the editor.
+export const editor = {
+  // what should happen when a translator pressed "Save"
+  // (edited a message).
+  onSave: async (message) => {
+    // the message could be committed.
+    return commit(message);
+    // or a pull request can be opened.
+    return openPullRequest(message);
+    // or something else can happen
+    return somethingElse(message);
+  },
+  // and more...
+};
+```
+
+### Conclusion
+
+The common denominator across all proposed components is a JavaScript [instead of JSON/YAML/TOML] config and an AST.
+
+The JS config solves the storage [of messages] and different syntaxes [to express human languages] problems by exposing functions to developers and allowing them to define how resources are read, parsed, serialized, and written to the filesystem. Furthermore, developers are empowered to adjust the business logic of inlang components to the needs of the project. Due to the complexity of sandboxing JS, the detailed design of the config will follow in RFC-003.
+
+The config itself requires an AST specification for developers to parse resources to and serialize from which is further consumed by all inlang components. TODO: THE CHOICE HAS NOT BEEN MADE YET.
+
+### Flowchart
+
+```mermaid
+flowchart LR
+    Applications
+    subgraph Packages
+        subgraph Providers
+                GitHub
+                Azure
+                GitLab
+                Other[etc...]
+        end
+        Providers-.-Host
+        git[git-js]
+        Host[git-provider-api]
+        validation
+        ast
+        config
+    end
+    subgraph Legend
+        direction LR
+        start1[ ] --->|depends on| stop1[ ]
```

Comment:
reverse direction of dependency

Reply:
Good remark. Removed the legend. The arrows are clear, the legend makes the relations unclear.